tests/unit/test_session.py

```diff
@@ -79,3 +84,119 @@ def test_used_scoped_temp_object():
+    finally:
+        os.remove(a_temp_file)
+
+
+def test_resolve_package_terms_not_accepted():
+    fake_connection = mock.create_autospec(ServerConnection)
+    fake_connection._conn = mock.Mock()
+    session = Session(fake_connection)
+
+    def get_information_schema_packages(table_name: str):
+        if table_name == "information_schema.packages":
+            result = MagicMock()
+            result.filter().group_by().agg()._internal_collect_with_tag.return_value = (
+                []
+            )
+            return result
+
+    def run_query(sql: str):
+        if sql == "select system$are_anaconda_terms_acknowledged()":
+            return [[False]]
+
+    session.table = MagicMock(name="session.table")
+    session.table.side_effect = get_information_schema_packages
+    session._run_query = MagicMock(name="session._run_query")
+    session._run_query.side_effect = run_query
+    with pytest.raises(
+        ValueError,
+        match="Cannot add package random_package_name because Anaconda terms must be accepted by ORGADMIN to use "
+        "Anaconda 3rd party packages. Please follow the instructions at "
+        "https://docs.snowflake.com/en/developer-guide/udf/python/udf-python-packages.html"
+        "#using-third-party-packages-from-anaconda.",
+    ):
+        session._resolve_packages(
+            ["random_package_name"], validate_package=True, include_pandas=False
+        )
+
+
+def test_write_pandas_wrong_table_type():
+    fake_connection = mock.create_autospec(ServerConnection)
+    fake_connection._conn = mock.Mock()
+    session = Session(fake_connection)
+    with pytest.raises(ValueError, match="Unsupported table type."):
+        session.write_pandas(
+            mock.create_autospec(pandas.DataFrame), table_name="t", table_type="aaa"
+        )
+
+
+def test_create_dataframe_empty_schema():
+    fake_connection = mock.create_autospec(ServerConnection)
+    fake_connection._conn = mock.Mock()
+    session = Session(fake_connection)
+    with pytest.raises(
+        ValueError,
+        match="The provided schema or inferred schema cannot be None or empty",
+    ):
+        session.create_dataframe([[1]], schema=StructType([]))
+
+
+def test_create_dataframe_wrong_type():
+    fake_connection = mock.create_autospec(ServerConnection)
+    fake_connection._conn = mock.Mock()
+    session = Session(fake_connection)
+    with pytest.raises(
+        TypeError, match=r"Cannot cast <class 'int'>\(1\) to <class 'str'>."
+    ):
+        session.create_dataframe([[1]], schema=StructType([StructField("a", str)]))
+
+
+def test_table_exists_invalid_table_name():
+    fake_connection = mock.create_autospec(ServerConnection)
+    fake_connection._conn = mock.Mock()
+    session = Session(fake_connection)
+    with pytest.raises(
+        SnowparkInvalidObjectNameException,
+        match="The object name 'a.b.c.d' is invalid.",
+    ):
+        session._table_exists("a.b.c.d")
+
+
+def test_explain_query_error():
```

Comment:
this is covered in https://github.com/snowflakedb/snowpark-python/pull/594/files